File: 416-partition-equal-subset-sum/partition-equal-subset-sum.java
class Solution {
    public boolean canPartition(int[] nums) 
    {
        int n=nums.length;
        int sum= Arrays.stream(nums).sum();
        if(sum%2==1)
        return false;
        sum/=2;
        boolean res[][]=new boolean[n][sum+1];
        for(int i=0;i<n;i++)
        res[i][0]=true;
        for(int i=0;i<n;i++)
        {
            for(int j=0;j<=sum;j++)
            {
                if(i==0)
                {
                    if(j-nums[i]==0)
                    res[i][j]=true;
                }
                else if (j>=nums[i])
                {
                   
                    res[i][j]= res[i-1][j] || res[i-1][j-nums[i]];
                }
                else
                res[i][j]=res[i-1][j];
            }
        }
    //   System.out.println(Arrays.deepToString(res));
        return res[n-1][sum];
    }
}

File: 416-partition-equal-subset-sum/partition-equal-subset-sum.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        s = sum(nums)
        if s%2==1:
            return False
        need = s//2
        n = len(nums)
        @cache
        def recur(idx,val):
            if val<0 or idx==n:
                return False
            if val==0:
                return True
            return recur(idx+1,val) or recur(idx+1,val-nums[idx])
        return  recur(0,need)